### ISSUE TITLE
FR: Provide a cause for the onDidChangeTextDocument

### ISSUE THREAD

**Issue by Chillee** [Microsoft/vscode#36939]
It would be very helpful for to provide a `kind`/cause attribute for the TextDocumentChangeEvent, similar to the one provided for `TextEditorSelectionChangeEvent`.

@rebornix

**Comment by jrieken** [comment]
Sorry, not possible with breaking the API. The event is specified as just the editor, no place for extra information.

**Comment by Chillee** [comment]
Sorry, I'm a bit confused what you mean by breaking the API. What do you mean by "the event is specified as just the editor"? It already contains both the document and the contentChanges, no? There also seems to an already be an identifier property on [`IIdentifiedSingleEditOperation`](https://github.com/Chillee/vscode/blob/quickPickHandler/src/vs/editor/common/model/editableTextModel.ts#L635), so I'd imagine you could piggy back on that.

For example, the important thing for me is that [`replace`](https://github.com/Chillee/vscode/blob/quickPickHandler/src/vs/editor/common/core/editOperation.ts#L31) is able to accept some kind of parameter that allows me to know where the change is coming from. So I'd imagine that I could just pass a string or something that is then propogated down the actual changeEvent  [here](https://github.com/Chillee/vscode/blob/quickPickHandler/src/vs/editor/common/model/editableTextModel.ts#L635)

Basically, my use case is that when I copy text over from Neovim, I have to replace the entire buffer. Usually, when there's a TextDocumentChangeEvent from VSCode, I sync that over to Neovim, but in this case, it can cause a race. 

As in, the typical passing of data is (Neovim -> VSCode), then due to the TextDocumentChangeEvent, we have to do (VSCode -> Neovim). We have no way of knowing that the TextDocumentChangeEvent we're capturing is due to us, so we have to recopy.

Now, if Neovim sends data twice, we have a race between handling the first TextDocumentChangeEvent and passing the data from Neovim -> VSCode the second time.

**Comment by eamodio** [comment]
@Chillee I think your title is wrong -- I think you meant `onDidChangeTextDocument` not `onDidChangeActiveTextEditor`

**Comment by Chillee** [comment]
@eamodio oh whups you're completely right.

**Comment by jrieken** [comment]
haha, I should read more than the title. A dupe of #10801 (closed) because we actually don't know what's caused the change

**Comment by Chillee** [comment]
@jrieken I don't actually need to know it in the general case, I just need to know it for text edits that my extension is causing. It's a similar case for the remote editing plugin that the issue linked has.

For this more limited case, is it not possible using an approach similar to what I mentioned in the comment? I could try hacking something if it's something that might be feasibly merged.
